### PR TITLE
Feat/add item

### PR DESCRIPTION
**Pull request by sivanharel**

### CODE REVIEW

**Comment by WixBuildServer** [comment]
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://bo.wix.com/pages/wix-ui-tpa.pr_423/

**Comment by sivanharel** [comment]
@shlomitc , @jonathanadler -

In this PR I used the AddItem component from WSR.
I would like to share with you all the difficulties and the open questions that we should think about.

**I am not done with the PR yet**, I still need to make some changes and I also need to talk with Michal to close the gaps from the design side. 
But for the meantime, I would like to share my thoughts :)

Since there is no fulfill that supports CSS variable transpiling for the IE11 yet, I used the stylable extend.

Here are some gaps that were raised while wrapping the WSR AddItem component:

1. **Inner Text component-** 
The WSR AddItem component warps the content that passes by the "children" prop with the Text component in order to add the design. Since we have a different design in TPA  it's required us to override all the text CSS in our wrapper component because it is not fit for the TPA design.
In addition, this Text component uses the Ellipsis component which has a lot of calculations that might be a problem from a performance perspective.
We might need a way to prevent the wrapping of the Text component when it used for TPA. But when thinking from WSR perspective, I am not sure it is something that we want to expose for other users because it will let our users to overriding the styling easily.

2. **Focus-** 
The WSR components add a blue outline and CSS for a18y that we need to remove in the TPA.It also uses the FocusableHOC which is not relevant. As I understood from Jonathan, the outline and the CSS for the accessibility is added in the viewer. In the TPA components, we are just adding some opacity to the elements.

3. **Bundle size-**
There is some logic that included in the AddItem WSR component that is not relevant for TPA. For example, the different themes, tooltip, and focus. In this case, it is not a lot of logic but it is something that might be more critical in other components since all this unneeded logic might affect the bundle size.

4. **Long CSS classes that might be changed-**
Since there are a lot of uses in stylable states (Pseudo-Classes) and the fact the WSR component is using inner components, it makes the CSS selector that we use inside the TPA components longer, and also it makes them be dependent on the inner implementation of WSR.

5. **Depending on the WSR inner implementation and No general CSS rules-**
It seems that we are not only changing color and also changing other CSS. Since in the WSR AddItem the CSS was divided into different states it required us overriding each one of them separately without the ability to define a global CSS rule. When working on this PR I was requested to go over the CSS file of the WSR AddItem and to override each state separately.

6. **Adding a state to the TPA AddItem component was a bit difficult-**
As requested in the TPA AddItem Spec, the component should have an error state. Since this state wasn't added to the WSR component, It required some CSS duplication in order to prevent overriding by the disabled state of WSR.

7. **Exposing WSR Driver functions-**
We have to decide which function is relevant to TPA.

8. **Testing hover and focus on our visual tests:**
Since we are making a lot of CSS overriding that might be affected by changes in the WSR AddItem component we must have the ability to add visual tests for hover and focus.
Currently, we are not sure how to do it. @jonathanadler  said that he will talk with Applitools about it.

**Review by shlomitc — CHANGES_REQUESTED** [review]

**Comment by shlomitc** [comment]
Thanks, @sivanharel - 
These are really important insights and I think they are very important for the theming/unified library we're tackling.
I'm sure @EgoziE will also be interested in it

Here are my comments:

1. Text component
    1. Need to override the CSS - that’s fine in my opinion, I wrote in a comment of a more declarative way to do so.
    2. Tooltip calculations - if it passes SSR tests, then it’s ok. Let us know if that’s not the case.
    3. Prevent wrapping - not sure I understand your concern, can you explain this one again?
2. Focus + Bundle - it’s an interesting point. Let’s gather all the “WSR Specific” concerns and see if we can refactor the original component in a way it won’t interfere.
3. See Above
4. Long CSS API - can you share the longest class override you have? And yes, it does mean that this API of WSR is a public API, so it shouldn’t change otherwise your component will break. To be on the safe side, we need to break the build in case of incorrect API usages, I will check this with Core3 guys + you should have visual tests for it so at least you will know if something breaks.
5. Styling states - You’re correct. When a component defines a state, your override will need to be specific as this state. I don’t know if you can do it better besides in WSR declare a variable per state (`dashesBorderColor`, `dashesDisabledBorderColor`). Not sure if I like this more…
6. New state requires more work - You’re correct, it’s again related to the specificity of the states. My only idea here is to **not** use stylable states but rather concatenate classes (like using `classNames(classes.dashes, classes.disabled)`). I didn’t test it but it might be better.
7. Exposing APIs - since you’re exporting a new component, that’s completely yours to decide the API. Let us know if you need a review.
8. Testing non-HTML states - If it isn’t doing a DOM change - then eyes-storybook visual tests can’t test it. You can do an `eyes.it` or similar tests (which will take a screenshot instead of the DOM structure). In my opinion - if it’s not easy to test, don’t waste the time on it (unless it could break often).

**Comment by EgoziE** [comment]
Nice work!
Your fourth and fifth comments are solvable by using css variables!
@shlomitc This is exactly why I think it's so important to use css variables.
They suppose to target direct pieces in the component, regardless the structure.

Following question, regarding your first point, does it affect Madefor in any way? is Madefor relevant here?

**Comment by shlomitc** [comment]
@EgoziE - you can also use stylable vars, no need for CSS vars if that's the case.
The downside, as we already noticed in the EditorX theme - you will need to define a variable that represents both the property and the state: e.g. - `dashesDisabledBorderColor`

File: src/components/AddItem/AddItem.st.css [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+/* Remove WSR focus, opacity is just for now */
+.root:not(:disabled):focus-visible:theme(dashes) {
+    box-shadow: none;
+    border: dashed 1px value(CurrentBorderColor);
+    opacity: 0.6;
+}
+
+.root::text Text {
```
[the remark this answers is not in the document]
Comment (sivanharel) [reply src/components/AddItem/AddItem.st.css:94]: You are right! I opened a PR in WSR.
https://github.com/wix/wix-style-react/pull/6067/files

File: src/components/AddItem/AddItem.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          enzymeAddItemTestkitFactory,
+          mount,
+          {
+            dataHookPropName: 'data-hook',
+          },
+        ),
+      ).toBe(true);
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (sivanharel) [reply src/components/AddItem/AddItem.spec.tsx:56]: I opened a separate task for it on Monday

File: src/components/AddItem/AddItem.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import * as React from 'react';
+import { TPAComponentProps } from '../../types';
+
+import { AddItem as WSRAddItem } from 'wix-style-react';
```
[the remark this answers is not in the document]
Comment (jonathanadler) [reply src/components/AddItem/AddItem.tsx:4]: I'm not sure I understand this part.
Do all `wix-style-react` consumers need to add this build step?

[the remark this answers is not in the document]
Comment (shlomitc) [reply src/components/AddItem/AddItem.tsx:4]: you must change the import path, otherwise the cherry-picking will not work (you will get the entire package in your bundle).

Regarding the build step - no, consumers don't need to do it in the project.
It's the library's responsibility to make sure it supports esm all the way (including in its node_modules).
Consumers will simply use webpack which knows how to handle commonjs/esm, given that the node modules are created correctly

[the remark this answers is not in the document]
Comment (jonathanadler) [reply src/components/AddItem/AddItem.tsx:4]: ok, so these are two different things, right?
1. import directly from `dist`
2. add the build step for the stylable files

[the remark this answers is not in the document]
Comment (jonathanadler) [reply src/components/AddItem/AddItem.tsx:4]: Maybe @EgoziE can give us a short explanation tomorrow on what this build step does?

File: src/components/AddItem/AddItem.driver.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  const WSRAddItemDriver = WSRAddItemUniDriverFactory(base, body);
+
+  return {
+    ...baseUniDriverFactory(base),
+
+    /**
+     * Hover on the element
+     * @returns {Promise<void>}
+     */
+    hover: async () => base.hover(),
```
[the remark this answers is not in the document]
Comment (sivanharel) [reply src/components/AddItem/AddItem.driver.ts:30]: Removed.
It was added for internal uses and it is not in use now.
I am simulating the hover in the e2e.

**Comment by sivanharel** [comment]
@shlomitc - I agree with you..
It's something that I added but then removed it since Jonathan told me that currently, all the SVG doesn't have the fill="currentColor" and in the components, we are just overriding the fill by CSS.

@jonathanadler - maybe I should add it again?

File: src/components/AddItem/AddItem.meta.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+AddItemMetadata.addSim({
+  title: 'render',
+  props: {
+    disabled: false,
+    hasError: false,
+    alignment: 'center',
+    iconSize: 'medium',
+    children: 'Add Item',
+  } as AddItemProps,
+});
```
Comment (jonathanadler) [inline src/components/AddItem/AddItem.meta.tsx:16]: Let's add a test for `disabled: true`

File: src/components/AddItem/AddItem.st.css [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+.wsrAddItemRoot:not(:rtl):direction(horizontal)::textWrapper {
+    max-width: 100%;
+    width: auto;
+    flex-basis: auto;
+    padding: 0 0 0 8px;
+}
+
+.wsrAddItemRoot:rtl:direction(horizontal)::textWrapper {
+    padding: 0 8px 0 0;
+}
```
Comment (jonathanadler) [inline src/components/AddItem/AddItem.st.css:129]: Just a general comment
Usually I prefer to avoid using the `rtl` prop in the stylesheet.

So for instance, instead of using a different padding for each direction, we usually do something like:

```css
.text::before {
  width: 8px;
  ...
}
```
and then `flex` just takes it from there.

In this specific case, it might not work, since we need to access WSR's state

File: src/components/AddItem/AddItem.tsx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      <TPAComponentsConsumer>
+        {({ rtl }) => (
+          <ThemeProvider
+            dataHook={this.props['data-hook']}
+            theme={theme(rootClassName)}
+          >
+            <WSRAddItem
+              className={st(
+                classes.wsrAddItemRoot,
+                { rtl, hasError, direction, noChildren },
```
Comment (jonathanadler) [inline src/components/AddItem/AddItem.tsx:98]: So as I wrote earlier, `rtl` might not be needed here.
But it's not critical at all.

File: src/components/AddItem/AddItem.meta.tsx (hunk 5 above)
Comment (jonathanadler) [reply src/components/AddItem/AddItem.meta.tsx:16]: Also since the text is optional, and the button can only have an icon, we should add a test with icon only.
But be aware, it will fail 😁

**Review by jonathanadler — COMMENTED** [review]
Amazing work 🚀 🚀 🚀 

I left some comments, but also 2 comments:
* regarding the icons, shouldn't they all be the same size (`24px` but with different padding)?
We need to consider this, so that working with icons will be much easier.
* We need to add a11y props: `aria-label` and `aria-labelledby`, they will need to be passed to the button in WSR 😬

**Comment by sivanharel** [comment]
> * regarding the icons, shouldn't they all be the same size (`24px` but with different padding)?
>   We need to consider this, so that working with icons will be much easier.

That's what Michal sent me.

**Comment by sivanharel** [comment]
> * We need to add a11y props: `aria-label` and `aria-labelledby`, they will need to be passed to the button in WSR 😬

You are right and it might be relevant for every WSR component that we are going to use in TPA..

**Comment by jonathanadler** [comment]
> > * We need to add a11y props: `aria-label` and `aria-labelledby`, they will need to be passed to the button in WSR 😬
> 
> You are right and it might be relevant for every WSR component that we are going to use in TPA..

Absolutely 💪

File: src/components/AddItem/AddItem.tsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                className,
+              )}
+              dataHook={DATA_HOOKS.ADD_ITEM}
+              disabled={disabled}
+              theme="dashes"
+              alignItems={alignment}
+              borderRadius="0"
+              // @ts-ignore
```
Comment (jonathanadler) [inline src/components/AddItem/AddItem.tsx:112]: Why is this needed here?

**Review by jonathanadler — APPROVED** [review]

File: src/components/AddItem/AddItem.tsx (hunk 8 above)
Comment (sivanharel) [reply src/components/AddItem/AddItem.tsx:112]: It is a good question.
Which I wanted to ask you and forgot...
I don't know why but if I remove it I am getting an error that I should pass a value that fits for the WSR sizes although I pass valid values

Comment (jonathanadler) [reply src/components/AddItem/AddItem.tsx:112]: Oh, I see, typescript 🤦‍♂️ 

So I guess the most correct thing to do is this:
```ts
import WSRAddItem, { AddItemSize as WSRAddItemSize }  from 'wix-style-react/dist/src/AddItem';
...
const WSR_SIZE_MAP: { [key as ICON_SIZE]: WSRAddItemSize } = {
  [ICON_SIZE.small]: 'tiny',
  [ICON_SIZE.medium]: 'small',
  [ICON_SIZE.large]: 'medium',
  [ICON_SIZE.xLarge]: 'large',
};
```

I'm not 100% sure, but I think this should work

**Comment by jonathanadler** [comment]
fixes #411
